handler/src/main/java/io/netty/handler/pcap/PCAPWriteHandler.java

```diff
@@ -0,0 +1,372 @@
+        ByteBuf udpBuf = ctx.alloc().buffer();
+
+        InetSocketAddress srcAddr = datagramPacket.sender();
+        InetSocketAddress dstAddr = datagramPacket.recipient();
+
+        logger.debug("Writing UDP Data of {} Bytes, Src Addr {}, Dst Addr {}", datagramPacket.content().readableBytes()
+                , srcAddr, dstAddr);
+
+        UDPPacket.writePacket(udpBuf,
+                datagramPacket.content(),
+                srcAddr.getPort(),
+                dstAddr.getPort());
+
+        completeUDPWrite(srcAddr, dstAddr, udpBuf, ctx.alloc());
+
+        udpBuf.release();
+    }
+
+    private void completeTCPWrite(InetSocketAddress srcAddr, InetSocketAddress dstAddr, ByteBuf tcpBuf,
+                                  ByteBufAllocator byteBufAllocator) throws IOException {
+
+        ByteBuf ipBuf = byteBufAllocator.buffer();
+        ByteBuf ethernetBuf = byteBufAllocator.buffer();
+
+        try {
+
+            if (srcAddr.getAddress() instanceof Inet4Address) {
+                IPPacket.writeTCPv4(ipBuf,
+                        tcpBuf,
+                        ipv4ToInt(srcAddr.getAddress()),
+                        ipv4ToInt(dstAddr.getAddress()));
+
+                EthernetPacket.writeIPv4(ethernetBuf, ipBuf);
+            } else {
+                IPPacket.writeTCPv6(ipBuf,
+                        tcpBuf,
+                        srcAddr.getAddress().getAddress(),
+                        dstAddr.getAddress().getAddress());
+
+                EthernetPacket.writeIPv6(ethernetBuf, ipBuf);
+            }
+
+            pCapWriter.writePacket(byteBufAllocator.buffer(), ethernetBuf);
+        } catch (IOException ex) {
+            tcpBuf.release();
+            throw ex;
+        } finally {
+            ipBuf.release();
+            ethernetBuf.release();
+        }
+    }
+
+    private void completeUDPWrite(InetSocketAddress srcAddr, InetSocketAddress dstAddr, ByteBuf udpBuf,
+                                  ByteBufAllocator byteBufAllocator) throws IOException {
+
+        ByteBuf ipBuf = byteBufAllocator.buffer();
+        ByteBuf ethernetBuf = byteBufAllocator.buffer();
+
+        try {
+
+            if (srcAddr.getAddress() instanceof Inet4Address) {
+                IPPacket.writeUDPv4(ipBuf,
+                        udpBuf,
+                        ipv4ToInt(srcAddr.getAddress()),
+                        ipv4ToInt(dstAddr.getAddress()));
+
+                EthernetPacket.writeIPv4(ethernetBuf, ipBuf);
+            } else {
+                IPPacket.writeUDPv6(ipBuf,
+                        udpBuf,
+                        srcAddr.getAddress().getAddress(),
+                        dstAddr.getAddress().getAddress());
+
+                EthernetPacket.writeIPv6(ethernetBuf, ipBuf);
+            }
+
+            pCapWriter.writePacket(byteBufAllocator.buffer(), ethernetBuf);
+        } catch (IOException ex) {
+            udpBuf.release();
```

Comment:
I would not do this here but let the caller do this as the caller is the one that allocated this buffer